File: DijkstrasAlgorithm/DijkstrasAlgorithm/Edge.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DijkstrasAlgorithm
{
    public class Edge
    {
        private readonly int _Cost;
        private readonly Vertex[] ends;

        public Edge(int cost, Vertex v, Vertex u)
        {
            _Cost = cost;
            ends = new[] { v, u };
        }

        public int Cost
        {
            get
            {
                return _Cost;
            }
        }

        public Vertex[] Ends
        {
            get
            {
                return ends;
            }
        }

        public bool ConnectedTo(Vertex v)
        {
            return Ends.Any(n => n == v);
        }

        public Vertex OtherNode(Vertex that)
        {
            return Ends.Where(v => v != that).Single();
        }

        public override string ToString()
        {
            return string.Format(
                "{0} ({1},{2})",
                Cost,
                this.Ends[0].Name,
                this.Ends[1].Name);
        }
    }
}

File: DijkstrasAlgorithm/dijkstra/edge.cpp
#include "edge.h"
#include "vertex.h"

using namespace std;

Edge::Edge( int cost, Vertex &a, Vertex &b ) : Cost( cost )
{
    ends[0] = &a;
    ends[1] = &b;
}

Edge::Edge( const Edge &e ) : Cost( e.Cost )
{
    ends[0] = e.ends[0];
    ends[1] = e.ends[1];
}

Edge::~Edge()
{
}

bool Edge::ConnectedTo( const Vertex &v ) const
{
    return find( ends, ends + 2, &v ) != ends + 2;
};

string Edge::ToString()
{
    stringstream t;
    t << Cost << " " << "(" << ends[0]->Id << ", " << ends[1]->Id << ")";
    return t.str();
}
